web/api/v1/api.go

```diff
@@ -805,6 +806,84 @@ func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
 	return apiAlerts
 }
 
+type metadataSet map[metadata]struct{}
+
+func (ms metadataSet) MarshalJSON() ([]byte, error) {
+	var res bytes.Buffer
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+	encoder := json.NewEncoder(&res)
+
+	res.WriteByte('[')
+	for metadata := range ms {
+		if err := encoder.Encode(metadata); err != nil {
+			return nil, err
+		}
+	}
+	res.WriteByte(']')
+
+	return res.Bytes(), nil
+}
+
+type metadata struct {
+	Type textparse.MetricType `json:"type"`
+	Help string               `json:"help"`
+	Unit string               `json:"unit"`
+}
+
+func (api *API) metricMetadata(r *http.Request) apiFuncResult {
+	metrics := map[string]metadataSet{}
+
+	limit := -1
+	if s := r.FormValue("limit"); s != "" {
+		var err error
+		if limit, err = strconv.Atoi(s); err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, errors.New("limit must be a number")}, nil, nil}
+		}
+	}
+
+	for _, tt := range api.targetRetriever.TargetsActive() {
+		for _, t := range tt {
+			for _, mm := range t.MetadataList() {
+				_, ok := metrics[mm.Metric]
+				m := metadata{Type: mm.Type, Help: mm.Help, Unit: mm.Unit}
+
+				// If the metric name is not in the map, we don't have any metadata yet.
+				if !ok {
+					metrics[mm.Metric] = make(map[metadata]struct{})
+					metrics[mm.Metric][m] = struct{}{}
+					continue
+				}
+
+				// If the metric name is in the map, find if the metadata is already in the set.
+				// This collapses the same metadata from different targets.
+				_, ok = metrics[mm.Metric][m]
+				if ok {
+					continue
+				}
+
+				// By now, we know the metadata is not in the set - add it.
+				metrics[mm.Metric][m] = struct{}{}
```

Comment:
You can write the whole block in a more compact way:

```Go
m := metadata{Type: mm.Type, Help: mm.Help, Unit: mm.Unit}
ms, ok := metrics[mm.Metric]
if !ok { // No metadataSet yet. Create it.
    ms = metadataSet{}
    metrics[mm.Metric] = ms
}
ms[m] = struct{}{}
```

Putting things into the metadataSet is idempotent and cheap. Just do it. Checking first if it is already there and then put it in only if needed is likely more expensive. And a lot more code.

Reply:
Done!